hmiddleware/dynoid/dynoid.go

```diff
@@ -0,0 +1,121 @@
+	populate := populateDynoID(audience, callback)
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			next.ServeHTTP(w, populate(r))
+		})
+	}
+}
+
+// Authorize populates the dyno identity blocks requests where the callback fails.
+func Authorize(audience string, callback dynoid.IssuerCallback) func(http.Handler) http.Handler {
+	populate := Populate(audience, callback)
+
+	return func(next http.Handler) http.Handler {
+		return populate(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if _, err := FromContext(r.Context()); err != nil {
+				w.WriteHeader(http.StatusForbidden)
+				fmt.Fprint(w, http.StatusText(http.StatusForbidden))
+				return
+			}
+
+			next.ServeHTTP(w, r)
+		}))
+	}
+}
+
+func AuthorizeSameSpace(audience string) func(http.Handler) http.Handler {
+	token, err := dynoid.ReadLocalToken(context.Background(), audience)
+	if err != nil {
+		return internalServerError("failed to load dyno-id")
+	}
+
+	return Authorize(audience, func(issuer string) error {
+		if issuer != token.IDToken.Issuer {
+			return &dynoid.UntrustedIssuerError{Issuer: issuer}
+		}
+
+		return nil
+	})
+}
+
+func internalServerError(format string, v ...any) func(http.Handler) http.Handler {
+	return func(http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintf(w, format, v...)
+		})
+	}
+}
+
+// AuthorizeSpace populates the dyno identity and blocks any requests that
+// aren't from one of the given spaces.
+func AuthorizeSpaces(audience, host string, spaces ...string) func(http.Handler) http.Handler {
+	return Authorize(audience, dynoid.AllowHerokuSpace(host, spaces...))
+}
+
+// FromContext fetches the Token from the context
+func FromContext(ctx context.Context) (*dynoid.Token, error) {
+	token, _ := ctx.Value(DynoIDKey).(*dynoid.Token)
+	err, _ := ctx.Value(DynoIDErrKey).(error)
+
+	return token, err
+}
+
+// AddToContext adds the Token to the given context
+func AddToContext(ctx context.Context, token *dynoid.Token, err error) context.Context {
+	ctx = context.WithValue(ctx, DynoIDKey, token)
+	ctx = context.WithValue(ctx, DynoIDErrKey, err)
+	return ctx
+}
+
+func populateDynoID(audience string, callback dynoid.IssuerCallback) func(*http.Request) *http.Request {
+	verifier := dynoid.NewWithCallback(audience, callback)
+
+	return func(r *http.Request) *http.Request {
+		ctx := r.Context()
+
+		rawToken := tokenFromHeader(r)
+		if rawToken == "" {
+			return r.WithContext(AddToContext(ctx, nil, ErrTokenMissing))
+		}
+
+		token, err := verifier.VerifyHeroku(r.Context(), rawToken)
+
+		return r.WithContext(AddToContext(ctx, token, err))
+	}
+}
+
+func tokenFromHeader(r *http.Request) string {
+	bearer := r.Header.Get("Authorization")
+	if len(bearer) > 7 && strings.ToUpper(bearer[0:6]) == "BEARER" {
```

Comment:
I suggest matching on the space too and making this case sensitive as recommended in [section 2.1 of RFC 7235](https://datatracker.ietf.org/doc/html/rfc7235#section-2.1):

> It uses a case-insensitive token as a means to identify the authentication scheme, followed by additional information necessary for achieving authentication via that scheme.

```suggestion
	if len(bearer) > 7 && bearer[:7] == "Bearer " {
```